#PBS -l walltime=20:00:00,nodes=1:ppn=4,mem=24gb

|----------------------------------------------------------------------------|
|This script was written *specifically* for the MMG434 course to analyze the |
|L. reuteri RNA-seq data generated by the MSU RTSF. This series of processes |
|will generate a folder called 'RNAseq_(currentdate)' in your home (~) direc-|
|tory. Within that folder a series of sub-folders are made to organize the   |
|data and analysis results. The general workflow is Trimmomatic -> FastQC -> |
|Bowtie -> HTSeq. After this has been run, you can download the map.sam files|
|and analyze them in your favorite differential gene expression software pac-|
|age (e.g. edgeR). To use this script, you will need to change the data loca-|
|tion and then save it (e.g. use nano or another text editor to make the cha-|
|nges). To run this script, follow the tutorial at mmg434.readthedocs.org.   |
|----------------------------------------------------------------------------|

#change to home directory on HPCC
cd ~

#check if the folder RNAseq_(current date) exists and if not, make the folder
NEWFOLDNAME=RNAseq
DATE=`date +%d-%m-%y`
NEWFOLD=${NEWFOLDNAME}_${DATE}

#if the folder already exists, make a new one with a _number added
n=1	#start with the number as 1

while [ -d ~/$NEWFOLD ]	#while the folder exists, do the commands below
do

	NEWFOLD=${NEWFOLDNAME}_${DATE}_${n}	#add the number to the folder name
	n=$((n+1))	#increment n by 1

done

mkdir $NEWFOLD
cd ~/$NEWFOLD

#make several folders within this RNAseq_(current date) folder to organize data
mkdir TrimmedData FastQCreports Bowtie HTSeq edgeR


#load the appropriate modules for all of the analysis steps
module load Trimmomatic/0.32
module load FastQC/0.11.2
module load bowtie/1.0.0
module load HTSeq/0.6.1

#Copy the data into the TrimmedData folder
#DEPENDING ON YOUR COMPARISON, YOU NEED TO CHANGE THE LBvsCOMM FOLDER LOCATION TO EITHER
#LBvsEHEC OR LBvsIndole
#IF YOU DO NOT CHANGE IT, YOU WILL RUN THE ANALYSIS FOR THE LBvsComm COMPARISON

cp /mnt/research/mmg434/LreuteriRNAseqData/RawData/LBvsComm/*.fastq.gz ~/$NEWFOLD/TrimmedData

#Do the trimming with Trimmomatic

cd ~/$NEWFOLD/TrimmedData
for untrimmedreads in ~/$NEWFOLD/TrimmedData/*
do
	FILENAME=`basename ${untrimmedreads%.*.*}`
	PREFIX=trimmed
	NEWTRIMFILE=${PREFIX}${FILENAME}
	java -jar $TRIM/trimmomatic SE -threads 4 $untrimmedreads $NEWTRIMFILE.fastq.gz ILLUMINACLIP:$TRIM/adapters/TruSeq3-SE.fa:2:30:10 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36
done

#Generate the FastQC reports

cp ~/$NEWFOLD/TrimmedData/trimmed*.fastq.gz ~/$NEWFOLD/FastQCreports
cd ~/$NEWFOLD/FastQCreports
for trimmedreads in ~/$NEWFOLD/FastQCreports/trimmed*.fastq.gz
do
	fastqc $trimmedreads
done
rm trimmed*.fastq.gz

#Decompress the reads files and move them to the Bowtie directory

cd ~/$NEWFOLD/TrimmedData
cp trimmed*.fastq.gz ~/$NEWFOLD/Bowtie

for compfiles in ~/$NEWFOLD/Bowtie/*.fastq.gz
do
	FILENAME=${compfiles%.*.*}
	gunzip -c $FILENAME.fastq.gz > $FILENAME.fastq
done

cd ~/$NEWFOLD/Bowtie
rm *.fastq.gz	#clean the .gz files

#Align the reads with Bowtie to the reference genome

cd ~/$NEWFOLD/Bowtie

bowtie-build /mnt/research/mmg434/LreuteriRNAseqData/RefGenomeFiles/trimmedlreuterijcm1112.fa trimmedlreuterijcm1112

for renametrim in ~/$NEWFOLD/Bowtie/*.fastq
do
	LESSPREFIX=`basename ${renametrim//trimmed/}`
	mv $renametrim ~/$NEWFOLD/Bowtie/$LESSPREFIX
done

for alignment in ~/$NEWFOLD/Bowtie/*.fastq
do	
	FILENAME=`basename ${alignment%.*}`
	PREFIX=align
	NEWALIGNFILE=${PREFIX}${FILENAME}
	bowtie -S -p 3 trimmedlreuterijcm1112 $FILENAME.fastq > $NEWALIGNFILE.sam
done

rm ~/$NEWFOLD/Bowtie/*.fastq #clean the .fastq files

#Count gene features with HTSeq

cp ~/$NEWFOLD/Bowtie/align*.sam ~/$NEWFOLD/HTSeq
cp /mnt/research/mmg434/LreuteriRNAseqData/RefGenomeFiles/alignlreuterijcm1112.gtf ~/$NEWFOLD/HTSeq
cd ~/$NEWFOLD/HTSeq

for renamealign in ~/$NEWFOLD/HTSeq/*.sam
do
	LESSPREFIX=`basename ${renamealign//align/}`
	mv $renamealign ~/$NEWFOLD/HTSeq/$LESSPREFIX
done

for counts in ~/$NEWFOLD/HTSeq/*.sam
do
	FILENAME=`basename ${counts%.*}`
	PREFIX=map
	NEWMAPFILE=${PREFIX}${FILENAME}
	htseq-count -m intersection-nonempty --stranded=reverse $FILENAME.sam alignlreuterijcm1112.gtf > $NEWMAPFILE.sam
done

rm ~/$NEWFOLD/HTSeq/LR*.sam	#clean the alignment files