<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Differential gene expression analysis with edgeR &mdash; MMG434 0.0.1 documentation</title>
    
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="MMG434 0.0.1 documentation" href="index.html" />
    <link rel="next" title="License" href="license.html" />
    <link rel="prev" title="Counting gene features with HTSeq" href="dayfour.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="license.html" title="License"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="dayfour.html" title="Counting gene features with HTSeq"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">MMG434 0.0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="differential-gene-expression-analysis-with-edger">
<span id="dayfive"></span><h1>Differential gene expression analysis with edgeR<a class="headerlink" href="#differential-gene-expression-analysis-with-edger" title="Permalink to this headline">¶</a></h1>
<p>Up to this point we have done several things: trimmed, aligned, QC&#8217;d the alignment, and counted reads that mapped to each gene. Now, we will finally move to the step where we will analyze the differential gene expression between the untreated and treated <em>L. reuteri</em> samples!</p>
<p>To do this, we have chosen to utilize an analysis package written in the R programming language called <a class="reference external" href="http://bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf">edgeR</a>. edgeR stands for differential expression analysis of digital gene expression data in R. This is a fantastic tool that is actively maintained (as seen by the date of the most recent user guide update) and fairly easy to use. Several diagnostic plots are produced throughout the analysis that provide meaningful information as to whether we can even perform differential gene expression between samples and if there are batch effects we have to deal with.</p>
<p>RNA-seq data does not typically assume a normal (Gaussian) distribution, so to glean which genes are changing in a statistically significant manner, we have to model the data slightly differently. EdgeR implements what is called a <a class="reference external" href="http://en.wikipedia.org/wiki/Negative_binomial_distribution#Related_distributions">negative binomial distribution</a>, sometimes referred to as a gamma-Poisson model. If you <em>really</em> enjoy statistics and would like to dig into the mathematical underpinnings of this software, see the references at the bottom of this page. If you are less interested in understanding the math behind all of this, here is the short summary: we need to examine the data to make sure they separate enough between treatments to determine differential gene expression and we <em>always</em> use a false-discovery rate correction to determine significance (even then, it&#8217;s worth looking at the fold-change differences to decide if it is &#8220;real&#8221;; though this is slightly more arbitrary).</p>
<p>Let&#8217;s get going!</p>
<p>We are going to do things in several steps today, with the goal being to be able to copy and paste as much of the code as we can.</p>
<ol class="arabic simple">
<li><a class="reference internal" href="#downloadedger"><em>Download and install edgeR</em></a></li>
<li><a class="reference internal" href="#readindata"><em>Read in the data to RStudio</em></a></li>
<li><a class="reference internal" href="#groupdata"><em>Group the data together</em></a></li>
<li><a class="reference internal" href="#filterdata"><em>Filter out low counts</em></a></li>
<li><a class="reference internal" href="#regroupandnormalize"><em>Regroup and normalize the libraries</em></a></li>
<li><a class="reference internal" href="#mdsplot"><em>See how samples separate by treatment</em></a></li>
<li><a class="reference internal" href="#diffexp"><em>Differential expression calculations</em></a></li>
<li><a class="reference internal" href="#plotres"><em>Plot the results</em></a></li>
</ol>
<div class="section" id="download-and-install-edger">
<span id="downloadedger"></span><h2>Download and install edgeR<a class="headerlink" href="#download-and-install-edger" title="Permalink to this headline">¶</a></h2>
<p>The first step is to open RStudio and download/install edgeR from the Bioconductor repository.</p>
<ol class="arabic simple">
<li>To install packages from Bioconductor, click next to the <strong>&gt;</strong> cursor. Type <strong>source(&#8220;http://bioconductor.org/biocLite.R&#8221;)</strong> and hit Enter/Return.</li>
</ol>
<img alt="BiocLite installer" class="align-center" src="_images/biocinstaller.jpg" />
<ol class="arabic simple" start="2">
<li>Then type <strong>biocLite(&#8220;edgeR&#8221;)</strong> and hit Enter/Return. RStudio should also install all the necessary dependencies as well.</li>
</ol>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">If RStudio asks &#8220;Update all/some/none? [a/s/n]:&#8221;, type <strong>a</strong> and then hit Enter/Return to update <strong>all</strong> of the outdated packages. It&#8217;s best to work with the most recent version of everything.</p>
</div>
</div>
<div class="section" id="read-in-the-data-to-rstudio">
<span id="readindata"></span><h2>Read in the data to RStudio<a class="headerlink" href="#read-in-the-data-to-rstudio" title="Permalink to this headline">¶</a></h2>
<p>The next step is to read in the data to RStudio.</p>
<ol class="arabic">
<li><p class="first">We are going to download the data from this site, as I&#8217;ve collated the data together for you from the HPCC. Please download all of the control samples and pick a treatment, and download all of those files to your desktop.</p>
<blockquote>
<div><ul class="simple">
<li><strong>LB controls</strong></li>
<li><a class="reference download internal" href="_downloads/LRWT1map.sam"><code class="xref download docutils literal"><span class="pre">LRWT1map.sam</span></code></a></li>
<li><a class="reference download internal" href="_downloads/LRWT2map.sam"><code class="xref download docutils literal"><span class="pre">LRWT2map.sam</span></code></a></li>
<li><a class="reference download internal" href="_downloads/LRWT3map.sam"><code class="xref download docutils literal"><span class="pre">LRWT3map.sam</span></code></a></li>
<li><a class="reference download internal" href="_downloads/LRWT4map.sam"><code class="xref download docutils literal"><span class="pre">LRWT4map.sam</span></code></a></li>
<li><strong>Indole treated</strong></li>
<li><a class="reference download internal" href="_downloads/LRindole1map.sam"><code class="xref download docutils literal"><span class="pre">LRindole1map.sam</span></code></a></li>
<li><a class="reference download internal" href="_downloads/LRindole2map.sam"><code class="xref download docutils literal"><span class="pre">LRindole2map.sam</span></code></a></li>
<li><a class="reference download internal" href="_downloads/LRindole3map.sam"><code class="xref download docutils literal"><span class="pre">LRindole3map.sam</span></code></a></li>
<li><a class="reference download internal" href="_downloads/LRindole4map.sam"><code class="xref download docutils literal"><span class="pre">LRindole4map.sam</span></code></a></li>
<li><a class="reference download internal" href="_downloads/LRindole5map.sam"><code class="xref download docutils literal"><span class="pre">LRindole5map.sam</span></code></a></li>
<li><a class="reference download internal" href="_downloads/LRindole6map.sam"><code class="xref download docutils literal"><span class="pre">LRindole6map.sam</span></code></a></li>
<li><strong>Commensal E.coli conditioned medium treated</strong></li>
<li><a class="reference download internal" href="_downloads/LRcomm1map.sam"><code class="xref download docutils literal"><span class="pre">LRcomm1map.sam</span></code></a></li>
<li><a class="reference download internal" href="_downloads/LRcomm2map.sam"><code class="xref download docutils literal"><span class="pre">LRcomm2map.sam</span></code></a></li>
<li><a class="reference download internal" href="_downloads/LRcomm3map.sam"><code class="xref download docutils literal"><span class="pre">LRcomm3map.sam</span></code></a></li>
<li><a class="reference download internal" href="_downloads/LRcomm4map.sam"><code class="xref download docutils literal"><span class="pre">LRcomm4map.sam</span></code></a></li>
<li><a class="reference download internal" href="_downloads/LRcomm5map.sam"><code class="xref download docutils literal"><span class="pre">LRcomm5map.sam</span></code></a></li>
<li><a class="reference download internal" href="_downloads/LRcomm6map.sam"><code class="xref download docutils literal"><span class="pre">LRcomm6map.sam</span></code></a></li>
<li><a class="reference download internal" href="_downloads/LRcomm7map.sam"><code class="xref download docutils literal"><span class="pre">LRcomm7map.sam</span></code></a></li>
<li><strong>EHEC conditioned medium treated</strong></li>
<li><a class="reference download internal" href="_downloads/LRehec1map.sam"><code class="xref download docutils literal"><span class="pre">LRehec1map.sam</span></code></a></li>
<li><a class="reference download internal" href="_downloads/LRehec2map.sam"><code class="xref download docutils literal"><span class="pre">LRehec2map.sam</span></code></a></li>
<li><a class="reference download internal" href="_downloads/LRehec3map.sam"><code class="xref download docutils literal"><span class="pre">LRehec3map.sam</span></code></a></li>
<li><a class="reference download internal" href="_downloads/LRehec4map.sam"><code class="xref download docutils literal"><span class="pre">LRehec4map.sam</span></code></a></li>
<li><a class="reference download internal" href="_downloads/LRehec5map.sam"><code class="xref download docutils literal"><span class="pre">LRehec5map.sam</span></code></a></li>
<li><a class="reference download internal" href="_downloads/LRehec6map.sam"><code class="xref download docutils literal"><span class="pre">LRehec6map.sam</span></code></a></li>
<li><a class="reference download internal" href="_downloads/LRehec7map.sam"><code class="xref download docutils literal"><span class="pre">LRehec7map.sam</span></code></a></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Now, we need to read the files into RStudio. To do this we need to create a variable for each file. I will give an example for each treatment that you should be able to copy and paste into RStudio.</p>
<blockquote>
<div><ul class="simple">
<li><strong>For LB control:</strong> wt1 = read.table(&#8220;~/Desktop/LRWT1map.sam&#8221;, row.names=1)</li>
<li><strong>For indole treated:</strong> in1 = read.table(&#8220;~/Desktop/LRindole1map.sam&#8221;, row.names=1)</li>
<li><strong>For E. coli commensal medium treated:</strong> co1 = read.table(&#8220;~/Desktop/LRcomm1map.sam&#8221;, row.names=1)</li>
<li><strong>For EHEC medium treated:</strong> eh1 = read.table(&#8220;~/Desktop/LRehec1map.sam&#8221;, row.names=1)</li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Repeat each of these commands for the respective treatment, making sure to change the variable name (e.g. wt1, in1, co1, eh1) each time (e.g. wt2 for LRWT2map.sam).</p>
</li>
<li><p class="first">Now we need to rename the sample column names. To do this, I will give an example for each treatment that you should be able to copy and paste into RStudio.</p>
<blockquote>
<div><ul class="simple">
<li><strong>For LB control:</strong> colnames(wt1) &lt;- &#8216;wt1&#8217;</li>
<li><strong>For indole treated:</strong> colnames(in1) &lt;- &#8216;in1&#8217;</li>
<li><strong>For E. coli commensal medium treated:</strong> colnames(co1) &lt;- &#8216;co1&#8217;</li>
<li><strong>For EHEC medium treated:</strong> colnames(eh1) &lt;- &#8216;eh1&#8217;</li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Repeat each of these commands for the respective treatment, making sure to change the variable name (e.g. wt1, in1, co1, eh1) each time (e.g. wt2 for LRWT2map.sam) and the new name (e.g. wt2).</p>
</li>
</ol>
</div>
<div class="section" id="group-the-data-together">
<span id="groupdata"></span><h2>Group the data together<a class="headerlink" href="#group-the-data-together" title="Permalink to this headline">¶</a></h2>
<p>So that we don&#8217;t have to work on each sample individually, we will put them all into a single variable.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">It is important that you put all of the same treatment type together in consecutive order (e.g. 1-7).</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">It is important that you put the control treatment (LB) as the first set of samples for our purposes. Otherwise, it&#8217;s simply important that you know what are your controls and treatments.</p>
</div>
<ol class="arabic">
<li><p class="first">To do this, copy and paste one of the following depending on which treatments you are comparing.</p>
<blockquote>
<div><ul class="simple">
<li><strong>For LB vs Indole:</strong> wtvin &lt;- cbind(wt1, wt2, wt3, wt4, in1, in2, in3, in4, in5, in6)</li>
<li><strong>For LB vs Commensal conditioned medium:</strong> wtvco &lt;- cbind(wt1, wt2, wt3, wt4, co1, co2, co3, co4, co5, co6, co7)</li>
<li><strong>For LB vs EHEC conditioned medium:</strong> wtveh &lt;- cbind(wt1, wt2, wt3, wt4, eh1, eh2, eh3, eh4, eh5, eh6, eh7)</li>
</ul>
</div></blockquote>
</li>
<li><p class="first">After you&#8217;ve copy and pasted this in, hit Enter/Return.</p>
</li>
<li><p class="first">Now, let&#8217;s group them by factor. Copy and paste the appropriate command based on what you are comparing.</p>
<blockquote>
<div><ul class="simple">
<li><strong>For LB vs Indole:</strong> group &lt;- factor(c(1,1,1,1,2,2,2,2,2,2))</li>
<li><strong>For LB vs Commensal conditioned medium:</strong> group &lt;- factor(c(1,1,1,1,2,2,2,2,2,2,2))</li>
<li><strong>For LB vs EHEC conditioned medium:</strong> group &lt;- factor(c(1,1,1,1,2,2,2,2,2,2,2))</li>
</ul>
</div></blockquote>
</li>
</ol>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This step simply allows us to separate each treatment from each other (i.e. control vs. treatment).</p>
</div>
</div>
<div class="section" id="filter-out-low-counts">
<span id="filterdata"></span><h2>Filter out low counts<a class="headerlink" href="#filter-out-low-counts" title="Permalink to this headline">¶</a></h2>
<p>It is important to filter out genes that have low read counts associated with them. This is because they will ultimately lead to a skewing of the data in subsequent steps of the analysis.</p>
<ol class="arabic">
<li><p class="first">We will generate an edgeR data structure called a DGEList. This will create the scaffold with which edgeR can access the data and do differential gene expression. Copy and paste the appropriate command for what you are comparing and then hit Enter/Return.</p>
<blockquote>
<div><ul class="simple">
<li><strong>For LB vs Indole:</strong> y &lt;- DGEList(counts=wtvin, group=group)</li>
<li><strong>For LB vs Commensal conditioned medium:</strong> y &lt;- DGEList(counts=wtvco, group=group)</li>
<li><strong>For LB vs EHEC conditioned medium:</strong> y &lt;- DGEList(counts=wtveh, group=group)</li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Now we will filter out genes that have counts greater than 2 CPM (counts per million reads mapped) in at least four samples.</p>
<blockquote>
<div><ul class="simple">
<li><strong>Type:</strong> keep &lt;- rowSums(cpm(y)&gt;2) &gt;= 4</li>
</ul>
</div></blockquote>
</li>
</ol>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">We may come back and refine this number when we have a look at our final scatter plot of the data.</p>
</div>
<ol class="arabic" start="3">
<li><p class="first">We will apply this filter we just made to our data set.</p>
<blockquote>
<div><ul class="simple">
<li><strong>Type:</strong> y &lt;- y[keep,]</li>
</ul>
</div></blockquote>
</li>
</ol>
</div>
<div class="section" id="regroup-and-normalize-the-libraries">
<span id="regroupandnormalize"></span><h2>Regroup and normalize the libraries<a class="headerlink" href="#regroup-and-normalize-the-libraries" title="Permalink to this headline">¶</a></h2>
<p>For this next set of steps, we will regroup our data now that we have applied this filter and normalize everything based on effective library size to prevent sequencing depth and library size from skewing the data.</p>
<ol class="arabic">
<li><p class="first">To regroup the data:</p>
<blockquote>
<div><ul class="simple">
<li><strong>Type:</strong> y$samples$lib.size &lt;- colSums(y$counts)</li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Now let&#8217;s normalize the data:</p>
<blockquote>
<div><ul class="simple">
<li><strong>Type:</strong> y &lt;- calcNormFactors(y)</li>
</ul>
</div></blockquote>
</li>
<li><p class="first">We can view what the scaling factor is by typing:</p>
<blockquote>
<div><ul class="simple">
<li><strong>Type:</strong> y$samples</li>
</ul>
</div></blockquote>
</li>
</ol>
</div>
<div class="section" id="see-how-samples-separate-by-treatment">
<span id="mdsplot"></span><h2>See how samples separate by treatment<a class="headerlink" href="#see-how-samples-separate-by-treatment" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="differential-expression-calculations">
<span id="diffexp"></span><h2>Differential expression calculations<a class="headerlink" href="#differential-expression-calculations" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="plot-the-results">
<span id="plotres"></span><h2>Plot the results<a class="headerlink" href="#plot-the-results" title="Permalink to this headline">¶</a></h2>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Differential gene expression analysis with edgeR</a><ul>
<li><a class="reference internal" href="#download-and-install-edger">Download and install edgeR</a></li>
<li><a class="reference internal" href="#read-in-the-data-to-rstudio">Read in the data to RStudio</a></li>
<li><a class="reference internal" href="#group-the-data-together">Group the data together</a></li>
<li><a class="reference internal" href="#filter-out-low-counts">Filter out low counts</a></li>
<li><a class="reference internal" href="#regroup-and-normalize-the-libraries">Regroup and normalize the libraries</a></li>
<li><a class="reference internal" href="#see-how-samples-separate-by-treatment">See how samples separate by treatment</a></li>
<li><a class="reference internal" href="#differential-expression-calculations">Differential expression calculations</a></li>
<li><a class="reference internal" href="#plot-the-results">Plot the results</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="dayfour.html"
                        title="previous chapter">Counting gene features with HTSeq</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="license.html"
                        title="next chapter">License</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/dayfive.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="license.html" title="License"
             >next</a> |</li>
        <li class="right" >
          <a href="dayfour.html" title="Counting gene features with HTSeq"
             >previous</a> |</li>
        <li><a href="index.html">MMG434 0.0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2015, Benjamin K Johnson.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3b2.
    </div>
  </body>
</html>