<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Install software and make sure everyone is on the same page &mdash; MMG434 0.0.1 documentation</title>
    
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="MMG434 0.0.1 documentation" href="index.html" />
    <link rel="next" title="RNA-seq background information, basic Linux/Unix commands, and Bowtie primer" href="daytwo.html" />
    <link rel="prev" title="Welcome to the MMG 434 RNA-seq tutorial" href="index.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="daytwo.html" title="RNA-seq background information, basic Linux/Unix commands, and Bowtie primer"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="index.html" title="Welcome to the MMG 434 RNA-seq tutorial"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">MMG434 0.0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="install-software-and-make-sure-everyone-is-on-the-same-page">
<span id="dayone"></span><h1>Install software and make sure everyone is on the same page<a class="headerlink" href="#install-software-and-make-sure-everyone-is-on-the-same-page" title="Permalink to this headline">¶</a></h1>
<p>Before moving forward to mapping reads to the genome, it is necessary to QC the reads and remove low quality reads and adapter sequences. Further we need the ability to perform differential gene expression between conditions. We have chosen to do this through several software packages. Below are some instructions on how to install them.</p>
<div class="section" id="software-to-install">
<h2>Software to install:<a class="headerlink" href="#software-to-install" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li><a class="reference internal" href="#trimmomatic"><em>Trimmomatic</em></a></li>
<li><a class="reference internal" href="#fastqc"><em>FastQC</em></a></li>
<li><a class="reference internal" href="#r"><em>R</em></a></li>
<li><a class="reference internal" href="#rstudio"><em>RStudio</em></a></li>
<li><a class="reference internal" href="#qualimap"><em>Qualimap</em></a></li>
</ol>
</div>
<div class="section" id="trimmomatic">
<span id="id1"></span><h2>Trimmomatic<a class="headerlink" href="#trimmomatic" title="Permalink to this headline">¶</a></h2>
<p>Trimmomatic is a commonly utilized tool for QC/trimming short reads and has been implemented in RNA-seq approaches in the literature and is the go to tool at the MSU sequencing facility (RTSF).</p>
<ol class="arabic simple">
<li>Trimmomatic can be downloaded from <a class="reference external" href="http://www.usadellab.org/cms/index.php?page=trimmomatic">here</a>.</li>
<li>Scroll down the page (which you can read at your leisure later if you are interested) to where it says &#8220;Download Trimmomatic&#8221;</li>
<li>Click on the link that says &#8220;binary&#8221;</li>
</ol>
<a class="reference internal image-reference" href="_images/trimbinary.jpg"><img alt="image showing area where to download the binary file for Trimmomatic" class="align-center" src="_images/trimbinary.jpg" style="width: 400px; height: 300px;" /></a>
<ol class="arabic simple" start="4">
<li>The file will end up in your downloads folder and will likely need to be unzipped as it is compressed. This can typically be accomplished just by double clicking on the file itself in your downloads folder.</li>
<li>Once this is done, move the folder to your desktop.</li>
</ol>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Trimmomatic runs on Java. Go <a class="reference external" href="http://www.java.com/en/">here</a> and click on &#8220;Download&#8221; and download/install the latest version of Java for your computer. <strong>Trimmomatic will not run unless you have Java installed</strong>.</p>
</div>
<ol class="arabic simple" start="6">
<li>Open the Trimmomatic folder on your desktop and navigate to the folder called <strong>adapters</strong>.</li>
<li>Drag and drop the file <strong>TruSeq3-SE.fa</strong> out of the adapters subfolder and into the main <strong>Trimmomatic</strong> folder. The list of files in the main folder will like be: an adapters folder, LICENSE, trimmomatic-0.32.jar, and TruSeq3-SE.fa.</li>
<li>You should be all set for installing this software!</li>
</ol>
</div>
<div class="section" id="fastqc">
<span id="id3"></span><h2>FastQC<a class="headerlink" href="#fastqc" title="Permalink to this headline">¶</a></h2>
<p>FastQC is an application that can be utilized to assess sequence data. It&#8217;s worth looking at the reads after trimming/before aligning to the genome.</p>
<ol class="arabic simple">
<li>FastQC can be downloaded <a class="reference external" href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/">here</a>.</li>
</ol>
<img alt="image showing area where to download FastQC" class="align-center" src="_images/fastqc.jpg" />
<img alt="image showing area where to download either Windows or Mac FastQC files" class="align-center" src="_images/fastqcfiledownload.jpg" />
<ol class="arabic simple" start="2">
<li>The software is well documented and quite powerful. Every statistic/assessment that it draws can be found in the <a class="reference external" href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/">FastQC documentation</a>. Further, you can watch a <a class="reference external" href="https://www.youtube.com/watch?v=bz93ReOv87Y">tutorial video</a> and view examples of <a class="reference external" href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/good_sequence_short_fastqc.html">good Illumina data</a> and <a class="reference external" href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/bad_sequence_fastqc.html">bad Illumina data</a>.</li>
<li>The file will end up in your downloads folder. If you are a <strong>Windows user</strong>, double click on the <strong>run_fastqc bat file</strong>. If you are a <strong>Mac</strong> user, double click on the <strong>fastqc_vX.XX.X.dmg file</strong> to run the application.</li>
</ol>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">FastQC runs on Java. Go <a class="reference external" href="http://www.java.com/en/">here</a> and click on &#8220;Download&#8221; and download/install the latest version of Java for your computer. <strong>FastQC will not run unless you have Java installed</strong>.</p>
</div>
<ol class="arabic simple" start="4">
<li>Please attempt to open the software to make sure that everything will at least open and run.</li>
<li>Congratulations! You&#8217;ve installed another piece of software for this module.</li>
</ol>
</div>
<div class="section" id="r">
<span id="id6"></span><h2>R<a class="headerlink" href="#r" title="Permalink to this headline">¶</a></h2>
<p>In order to do differential gene expression to compare treatments and identify what is changing at the transcript level, we need to install the statistical programming language R and a really neat interface to work in R called RStudio.</p>
<ol class="arabic simple">
<li>R can be downloaded from <a class="reference external" href="http://cran.mtu.edu/">here</a>.</li>
<li>Click on the appropriate link for your operating system (Linux, Mac OS X, or Windows).</li>
</ol>
<img alt="image showing the list of platforms that R can be downloaded/installed for" class="align-center" src="_images/Rplatformselection.jpg" />
<ol class="arabic simple" start="3">
<li>Then, click on the latest version of the software. This will initiate the download. <strong>Windows users</strong> unless you know you already have R installed, click on the <strong>install R for the first time</strong> link (see below). <strong>Mac users</strong> you have two options based on what flavor of OS X you have. If you are not sure what version you have, look at the screen shots below.</li>
</ol>
<p><strong>Windows users:</strong></p>
<img alt="R for Windows" class="align-center" src="_images/RforWindows.jpg" />
<p><strong>Mac users (to determine OS X version):</strong></p>
<img alt="About this Mac navigation" class="align-center" src="_images/applebutton.jpg" />
<img alt="Version of Mac OS X" class="align-center" src="_images/appleversion.jpg" />
<p><strong>Mac users (version of R to download):</strong></p>
<img alt="R for Mac based on OS X version" class="align-center" src="_images/RforMac.jpg" />
<ol class="arabic simple" start="4">
<li>After the download finishes, double-click on the file and follow the instructions to install the software.</li>
<li>Congratulations! You&#8217;ve installed yet another piece of software for this module.</li>
</ol>
</div>
<div class="section" id="rstudio">
<span id="id8"></span><h2>RStudio<a class="headerlink" href="#rstudio" title="Permalink to this headline">¶</a></h2>
<p>RStudio is a fantastic interface to work in R. R does have a graphical user interface (GUI) that you can download and use, however I find RStudio much more intuitive/easier to use.</p>
<ol class="arabic simple">
<li>RStudio can be downloaded <a class="reference external" href="http://www.rstudio.com/products/RStudio/#Desk">here</a>.</li>
<li>We want to download and install the open-source version of RStudio for the desktop.</li>
</ol>
<img alt="RStudio for desktop download button" class="align-center" src="_images/RStudio.jpg" />
<ol class="arabic simple" start="3">
<li>Double click the file after the software has finished downloading and follow the instructions to install the software.</li>
<li>That&#8217;s it!</li>
</ol>
</div>
<div class="section" id="qualimap">
<span id="id10"></span><h2>Qualimap<a class="headerlink" href="#qualimap" title="Permalink to this headline">¶</a></h2>
<p>The last piece of software that is nice to have is a software suite that works with Java and R to generate PDF documents summarizing the data post-alignment to the genome.</p>
<ol class="arabic simple">
<li>Qualimap can be downloaded <a class="reference external" href="http://qualimap.bioinfo.cipf.es/">here</a>.</li>
<li>Click on the .zip file (as seen below)</li>
</ol>
<img alt="Qualimap download" class="align-center" src="_images/qualimapdownload.jpg" />
<ol class="arabic simple" start="3">
<li>The file will end up in your downloads folder and will likely need to be unzipped as it is compressed. This can typically be accomplished just by double clicking on the file itself in your downloads folder.</li>
<li>Once this is done, move the folder to your desktop.</li>
<li>Before we can run this software suite, we will need to install a few packages in R: optparse (from CRAN), NOISeq, Repitools, Rsamtools, GenomicFeatures, rtracklayer (all available from Bioconductor).</li>
<li>To do this, open RStudio, and click on the <strong>packages</strong> tab and then the <strong>Install</strong> button.</li>
</ol>
<img alt="Install new packages in RStudio" class="align-center" src="_images/RStudioinstallpackages.jpg" />
<ol class="arabic simple" start="7">
<li>Type <strong>optparse</strong> into the <strong>Packages (separate multiple with space or comma):</strong> field. Then click <strong>Install</strong>. RStudio should do the rest.</li>
<li>To install the packages from Bioconductor, click next to the <strong>&gt;</strong> cursor. Type <strong>source(&#8220;http://bioconductor.org/biocLite.R&#8221;)</strong> and hit Enter/Return.</li>
</ol>
<img alt="BiocLite installer" class="align-center" src="_images/biocinstaller.jpg" />
<ol class="arabic simple" start="9">
<li>Then type <strong>biocLite(&#8220;NOISeq&#8221;)</strong> and hit Enter/Return. Repeat for Repitools, Rsamtools, GenomicFeatures, and rtracklayer.</li>
<li>That&#8217;s all the software we need to install for now!</li>
</ol>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Install software and make sure everyone is on the same page</a><ul>
<li><a class="reference internal" href="#software-to-install">Software to install:</a></li>
<li><a class="reference internal" href="#trimmomatic">Trimmomatic</a></li>
<li><a class="reference internal" href="#fastqc">FastQC</a></li>
<li><a class="reference internal" href="#r">R</a></li>
<li><a class="reference internal" href="#rstudio">RStudio</a></li>
<li><a class="reference internal" href="#qualimap">Qualimap</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="index.html"
                        title="previous chapter">Welcome to the MMG 434 RNA-seq tutorial</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="daytwo.html"
                        title="next chapter">RNA-seq background information, basic Linux/Unix commands, and Bowtie primer</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/dayone.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="daytwo.html" title="RNA-seq background information, basic Linux/Unix commands, and Bowtie primer"
             >next</a> |</li>
        <li class="right" >
          <a href="index.html" title="Welcome to the MMG 434 RNA-seq tutorial"
             >previous</a> |</li>
        <li><a href="index.html">MMG434 0.0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2015, Benjamin K Johnson.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3b2.
    </div>
  </body>
</html>